### ISSUE TITLE
Revisit potential slice index bottleneck on yugabyte

### ISSUE THREAD

**Issue by jrudolph** [akka/akka-persistence-r2dbc#174]
Running performance tests on yugabyte we observed that the slice index seems to be a bottleneck, i.e. many or most requests at a time going to a single tablet.

This might be due to the fact, that we currently define the index to use ["range sharding"](https://docs.yugabyte.com/latest/architecture/docdb-sharding/sharding/) in this definition:

```sql
CREATE INDEX IF NOT EXISTS event_journal_slice_idx ON event_journal(entity_type ASC, slice ASC, db_timestamp ASC)
  INCLUDE (seq_nr, deleted);
```

Since `entity_type` and `slice` are usually of low cardinality it should be possible to change these at least to use `hash sharding`, instead, which might distribute the load better.

We should revisit what we require from the index both in terms of read and write performance and structure the keys / sharding accordingly.

**Comment by patriknw** [comment]
The reason is that it is a range index, which seems to default to one tablet. It is possible to define explicit split points for range sharded tables, and I that seems to work (I tried) for an index like this also (even though it's not documented for indexes). With the slices we know range and should define the split points if we need more than one tablet for the index.

**Comment by jrudolph** [comment]
> which seems to default to one tablet

Good point, that seems to be supported by https://docs.yugabyte.com/latest/architecture/docdb-sharding/tablet-splitting/#range-sharded-tables

I have another guess, that might be that sharding is determined based on only the first column in the index and since we used the same `entity_type` all over, we might have stayed in a single bucket.

You already suggested that we use `(slice, entity_type) HASH` instead, which sounds like it should work well for writing.

For reading, I guess this would mean that data for a given time interval is split over several tablets. That sounds fine, since it should be computationally simple to merge results from multiple tablets into a single stream ordered by `db_timestamp` (basically an n-way merge sort over already sorted partitions).

**Comment by patriknw** [comment]
It's not the order of the columns. I have it already like that in Akka Serverless, and this is what it looks like right now (singel tablet):
```
db1 | journal_slice_idx | 00004000000030008000000000004017 | cb8975249ad142ad8c298f487f66703c | range: [(<start>, <end>))
```

**Comment by jrudolph** [comment]
> It's not the order of the columns. I have it already like that in Akka Serverless, and this is what it looks like right now (singel tablet):

Ok, not sure though what the output means :)

**Comment by patriknw** [comment]
it was only one tablet for the index, and the range is covering everything (`range: [(<start>, <end>)`)

**Comment by patriknw** [comment]
I can confirm that the `journal_slice_idx` is a bottleneck if we don't split it. I noticed that one tablet server were handling more writes than others and I couldn't increase throughput even though db nodes were not much loaded.

<img width="1914" alt="Screenshot 2021-12-17 at 16 20 08" src="https://user-images.githubusercontent.com/336161/146570139-eff56e32-4e44-499c-917d-7ce5c39d8b1a.png">

The tablet for the journal_slice_idx had it's leader on that node.
<img width="1909" alt="Screenshot 2021-12-17 at 16 26 51" src="https://user-images.githubusercontent.com/336161/146570257-476f1d42-285c-4918-be24-7ba3c1aade55.png">

Removed the index and throughput doubled, from 2,700 rps to 6,600 rps.
<img width="1927" alt="Screenshot 2021-12-17 at 16 29 39" src="https://user-images.githubusercontent.com/336161/146570474-38957b31-126f-4d72-8530-4daf1d317915.png">
